modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cluster/ClientClusterGroupProjection.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.platform.client.cluster;
+
+import org.apache.ignite.binary.BinaryRawReader;
+import org.apache.ignite.cluster.ClusterGroup;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Client cluster group projection representation.
+ * Decodes a remote client's projection request.
+ */
+public class ClientClusterGroupProjection {
+    /** */
+    private static final short ATTRIBUTES = 1;
+    /** */
+    private static final short SERVER_NODE = 2;
+
+    /** Filter value mappings. */
+    private final Map<Short, Object> mappings;
+
+    /**
+     * Constructor.
+     *
+     * @param mappings Mappings.
+     */
+    private ClientClusterGroupProjection(Map<Short, Object> mappings){
+        this.mappings = mappings;
+    }
+
+    /**
+     * Reads projection from a stream.
+     * @param reader Reader.
+     * @return Projection.
+     */
+    public static ClientClusterGroupProjection read(BinaryRawReader reader)
+    {
+        if(!reader.readBoolean())
+            return new ClientClusterGroupProjection(new HashMap<>());
+
+        int cnt = reader.readInt();
+        Map<Short, Object> mappings = new HashMap<>(cnt);
+        for(int i = 0; i < cnt; i++){
+            short code = reader.readShort();
+            switch (code){
+                case ATTRIBUTES:{
+                    HashMap<String, String> attrs = new HashMap<>();
+                    int attrsCnt = reader.readInt();
+                    for(int j = 0; j < attrsCnt; j++){
+                        String attrName = reader.readString();
+                        String attrVal = reader.readString();
+                        attrs.put(attrName, attrVal);
+                    }
+                    mappings.put(ATTRIBUTES, attrs);
```

Comment:
If there is multiple `ATTRIBUTES` entries, we'll silently overwrite the previous one in the map.
We should either support this scenario properly or throw an error.

I lean towards supporting this - just replace Map with List<Pair>. Also avoids unnecessary HashMap overhead - we don't need hashing, just iteration.